### ISSUE TITLE
Provide per-app screenshots for AppData

### ISSUE THREAD

**Issue by jiri-janousek** [tiliado/nuvolasdk#5]
~~**Blocked by:** tiliado/nuvolasdk#8~~
 
Currently, Nuvola uses screenshots of Deezer for all scripts, which is weird at least, e.g. you look for "Spotify" in GNOME Software and it shows an app with the screenshots of Deezer.

Since I cannot expect from script maintainers to create and periodically update their screenshots with Nuvola's desktop integration features, it would be better to create generic screenshots of the features and then replace the generic web view with a screenshot of the actual web view of a particular app.

  - First, it is necessary to agree on the dimensions of the web view screenshots that are suitable for most apps. Nuvola should provide a Web View sidebar as a part of developer's tools. The sidebar will be used to retrieve current dimensions as well as to resize the web view to dimensions entered by the developer. 

    Each maintainer should provide **the minimal size** when all playback controls are visible and the web page is usable, i.e., there is no clipping of the web page and no important elements are hidden behind a scrollbar.

  - Once we agree on the dimensions, the maintainers will use the Web View sidebar to set the dimensions, take a snapshot of the web view and save it as `src/webview.png`.
  - Nuvola SDK will combine the base screenshot and the web view snapshot into a final screenshot, which will be hosted on https://nuvola.tiliado.eu and referenced in AppData XML.
 - Script maintainer may reference the screenshots in README.md.

Tasks:
- [x] Nuvola adds Web View sidebar to get/set web view dimensions and take a snapshot.
- [x] Maintainers are asked to provide minimal dimensions for their web apps.
- [x] Suitable dimensions are selected and maintainers are asked to provide web view snapshots.
- [x] Nuvola SDK combines base screenshots with web view snapshots.
- [x] The final screenshots are pushed to https://nuvola.tiliado.eu
- [x] Nuvola SDK references the screenshots in AppData XML.
- [ ] Maintainers are asked to reference the final screenshots in README.

**Comment by jiri-janousek** [comment]
Web View Sidebar
![nuvola_webview_sidebar](https://user-images.githubusercontent.com/853706/39189999-f5260764-47d3-11e8-91d0-0427f4c69c9c.png)

**Comment by jiri-janousek** [comment]
* "All screenshots should have a 16:9 aspect ratio, and should have a width that is no smaller than 620px..." -- [AppStream ⁠Chapter 4. Metadata Quickstart](https://www.freedesktop.org/software/appstream/docs/chap-Quickstart.html)
* 16:9 = 1280×768, 1600×900
* elementaryOs: dock  90 px, top bar 35 px, notifications 320 px from the right side, appindicator menu 385 px from right, mpris sound applet 389 from right => 1280×768 → 890×643 for window
* GNOME: top bar 35px, appindicator menu 238 px from right, MPRIS 385px from right, dock 75 px from left => 1280×768 → 820×733 for window
* Unity: left dock 76 px, quick list 175 px, top bar 35, MPRIS 335 px, ~~HUD 300  px~~ => 1280×768 → 695×733 for window

**Comment by jiri-janousek** [comment]
- The most common display resolutions of web users are HD 1366×768 and FHD 1920×1080 ([source](https://en.wikipedia.org/wiki/Display_resolution)).
- Ideally, the final screenshot should fit the web browser window without downscaling. Unlikely on HD...
- GNOME Software uses approx. 806×423 px area for screenshots.
- Large screenshot, large file.
- Therefore, 1600×900 is too large. I'd go with 1280×768 for the final screenshots and the integration/desktop widgets will overlap the web view partially.

**Comment by jiri-janousek** [comment]
A few examples of my scripts:

* Deezer 991×600 (obviously, the minimal content width was set to show ads fully) ![deezer](https://user-images.githubusercontent.com/853706/39299567-4d906798-4949-11e8-95a2-c35ec0c4c411.png)

* Spotify 736×600 (ad clipped)
![spotify](https://user-images.githubusercontent.com/853706/39299883-2b0521c2-494a-11e8-8f4f-e8329f43a3a6.png)

* Google Play Music 946×600 (with the scrollbar clipped) ![google_play_music](https://user-images.githubusercontent.com/853706/39300044-b3831522-494a-11e8-99b1-f419c11c3d18.png)

**Comment by jiri-janousek** [comment]
An additional guideline for web view snapshots: Try to remove annoing elements (ads, nag banners): 
![web_view_snapshot_remove_ads](https://user-images.githubusercontent.com/853706/39300898-94cd6ba2-494d-11e8-9048-57f2fd8eae50.png)


Deezer 2 983×600 
![deezer2](https://user-images.githubusercontent.com/853706/39300751-0ab43248-494d-11e8-8804-6b01846b92d5.png)

**Comment by jiri-janousek** [comment]
Hello @a-m-s, could you jump in? What are the minimum dimensions of the web view for your scripts? Please use the Web View sidebar from eu.tiliado.NuvolaAdk//master (that of stable suffers from a GTK+ bug and hangs).

**Comment by a-m-s** [comment]
I'm assuming you want a player view, rather than just the home page?

Jupiter broadcasting has a responsive design that changes as it gets smaller, so "minimal" is a tricky business. 1280x768 is fine. 991x600 is not, but 736x600 is fine again, although the navigation bar has now shrunk to a hamburger menu.
1280x768
![jb1280x768](https://user-images.githubusercontent.com/5088295/39302923-235d9290-494c-11e8-811b-8b22f92ba116.png)
991x600
![jb991x600](https://user-images.githubusercontent.com/5088295/39302940-3836d118-494c-11e8-8db5-26bd466e1433.png)
736x600
![jp736x600](https://user-images.githubusercontent.com/5088295/39302969-57b75116-494c-11e8-8795-a1598f39bee3.png)


BBC iPlayer is fine at any size, although more and more navigation items move to the "more" drop down. 1280x768 is good. It doesn't have visible player controls unless you mouse-over anyway.
1280x768
![bbc1280x768](https://user-images.githubusercontent.com/5088295/39302805-cfab81c0-494b-11e8-9f57-29ee8af0f1fb.png)
991x600
![bbc991x600](https://user-images.githubusercontent.com/5088295/39302849-f13ea2f4-494b-11e8-8b74-3e4ecc5382e2.png)

Amazon needs a horizontal scroll bar below 1020 pixels wide. The vertical width varies according to content, but clipping it would not be a problem.
1280x768
![amazon1280x768](https://user-images.githubusercontent.com/5088295/39302764-b103aa36-494b-11e8-9a8a-bf2f4e92b4bc.png)

**Comment by jiri-janousek** [comment]
> Jupiter broadcasting has a responsive design that changes as it gets smaller, so "minimal" is a tricky business. 1280x768 is fine. 991x600 is not, [...]

How about 1080×650?

> BBC iPlayer is fine at any size, although more and more navigation items move to the "more" drop down. 1280x768 is good. It doesn't have visible player controls unless you mouse-over anyway.
1280x768

Can be the player controls captured or do they fade away immediately?

---

It's time to spin up my virtual machines and test following: 

* The virtual display size: 1280×768
* The web view width: 1080 px => 200 px left, enough space for a vertical dock or for Nuvola's sidebar
* The web view height: 650 px => 118 px left, enough space for a horizontal dock, a system bar and a title bar

**Comment by a-m-s** [comment]
1080x650 is not so great. The controls are no longer visible.
![jb1080x650](https://user-images.githubusercontent.com/5088295/39310094-75087992-4961-11e8-9458-a6d8efdab34a.png)

The BBC controls fade very quickly. I could capture it with another snapshot tool. Of course, if the controls are visible then it's playing a video and the image in the viewer is likely to be less inspiring.

**Comment by jiri-janousek** [comment]
> [me] 16:9 = 1280×768, 1600×900

Well, 1280×720 is 16:9, not 1280×768 :-/

> 1080x650 is not so great. The controls are no longer visible.

Which is the best width < 1080 px for 650 px height?

> The BBC controls fade very quickly. I could capture it with another snapshot tool. Of course, if the controls are visible then it's playing a video and the image in the viewer is likely to be less inspiring.

Fair point, use whathever screen is the most representative for the web app. Btw, I've added an optional delay to the snapshot tool. I think I will need it for YouTube.

**Comment by a-m-s** [comment]
The website changes layout at 960px, so the "best" is 959x650. 1280x720 is fine too.

**Comment by jiri-janousek** [comment]
* This is an elementaryOS Loki base screenshot with 1080×600 transparent web view area: ![loki-bg](https://user-images.githubusercontent.com/853706/39676316-cea068e8-5168-11e8-9888-a644f0826af6.png)
* The same for Ubuntu 16.04 with Unity: ![unity-bg](https://user-images.githubusercontent.com/853706/39676747-fc53e6e6-516f-11e8-961d-a8fd97515844.png)
* So we really cannot have 1080×650 but only 1080×600 :-(

**Comment by jiri-janousek** [comment]
```
 ~/dev/repo/nuvola/apps/nuvola-app-deezer $ make all
python3 -m nuvolasdk create-addondata  -o nuvola-app-deezer.metainfo.xml -m metadata.json
python3 -m nuvolasdk create-screenshots -o screenshots -i src/webview.png
'/home/fenryxo/dev/repo/nuvola/nuvolasdk/nuvolasdk/data/screenshots/pantheon.png' → '/home/fenryxo/dev/repo/nuvola/apps/nuvola-app-deezer/screenshots/pantheon.png'
'/home/fenryxo/dev/repo/nuvola/nuvolasdk/nuvolasdk/data/screenshots/unity.png' → '/home/fenryxo/dev/repo/nuvola/apps/nuvola-app-deezer/screenshots/unity.png'
Timestamp file: '/home/fenryxo/dev/repo/nuvola/apps/nuvola-app-deezer/screenshots/timestamp.txt'
```
![pantheon](https://user-images.githubusercontent.com/853706/39677004-71f19dfe-5174-11e8-9fdd-8c0ef35c58f0.png)

It looks ok on white background, but in reality, there are semitransparent pixels! Need to read more Pillow doc.

**Comment by jiri-janousek** [comment]
Fixed:
![pantheon](https://user-images.githubusercontent.com/853706/39677318-2ce7d50c-5179-11e8-942b-1a9eef78eb21.png)

**Comment by jiri-janousek** [comment]
Hello @a-m-s, could you try that with your script? Save a screenshot as `src/webview.png`, install the latest Nuvola ADK and run `./configure; make all`. The web view area is 1080×600. If the screenshot is narrower than 1080, it is centered and the rest is filled with a gray color.

The base screenshots are not final (e.g. the Unity one has a scrollbar that should be removed eventually), but they are good enough for testing.

**Comment by a-m-s** [comment]
It seems to work ok, although the base image still says "Deezer".
![unity](https://user-images.githubusercontent.com/5088295/39863685-47b19c6a-543f-11e8-8e53-ceac71a04522.png)

**Comment by jiri-janousek** [comment]
It will say "Nuvola Player" or something like that in the final base screenshot. I've been working on a player demo script to be used to create base screen shots.

**Comment by a-m-s** [comment]
Do I need to do anything else now?

**Comment by jiri-janousek** [comment]
You can commit the web view snapshots. I'll then use them when the final base screenshots are ready.

**Comment by jiri-janousek** [comment]
> I've been working on a player demo script to be used to create base screen shots.

→ #8

**Comment by jiri-janousek** [comment]
:tada: All scripts have a web view screenshot.

**Comment by jiri-janousek** [comment]
New base screenshots show "Nuvola Player" instead of "Deezer".

**Comment by jiri-janousek** [comment]
- Screenshots are built with Circle CI: https://circleci.com/gh/tiliado
- I just need to download them and publish to https://nuvola.tiliado.eu/screenshots/apps

**Comment by jiri-janousek** [comment]
[Fixed in Nuvola 4.14](https://medium.com/nuvola-news/nuvola-4-14-released-533f1677ab5b)
